server/src/main/java/org/folio/shared/index/storage/Storage.java

```diff
@@ -486,60 +361,38 @@ static JsonObject handleRecord(Row row) {
+              JsonArray values = new JsonArray();
+              rowSet.forEach(row -> values.add(row.getString("match_value")));
+              object.put("matchValues", values);
+              return object;
+            }));
+  }
+
+  private static String postgresString(String s) {
+    return "'" + s + "'";
   }
 
   /**
    * return all clusters as streaming result.
    * @param ctx routing context
-   * @param matchKeyConfigId match ke config to use
+   * @param matchKeyId match ke config to use
    * @return async result
    */
-  public Future<Void> getAllClusters(RoutingContext ctx, String matchKeyConfigId) {
-    String q = "SELECT DISTINCT ON(cluster_id) cluster_id FROM " + clusterRecordTable
-        + " WHERE match_key_config_id = '" + matchKeyConfigId + "'";
-
-    return pool.getConnection().compose(connection ->
-        connection.prepare(q)
-            .onFailure(pq -> connection.close())
-            .compose(pq -> connection.begin().compose(tx -> {
-              HttpServerResponse response = ctx.response();
-              response.setChunked(true);
-              response.setStatusCode(200);
-              response.putHeader("Content-Type", "application/json");
-              RowStream<Row> stream = pq.createStream(sqlStreamFetchSize);
-              response.write("{ \"items\" : [\n");
-              AtomicInteger cnt = new AtomicInteger();
-              stream.handler(row -> {
-                if (cnt.incrementAndGet() > 1) {
-                  response.write(",\n");
-                }
-                UUID clusterId = row.getUUID("cluster_id");
-                stream.pause();
-                getClusterById(connection, clusterId)
-                    .onFailure(e -> {
-                      log.error(e.getMessage(), e);
-                      stream.close();
-                    })
-                    .onSuccess(obj -> {
-                      response.write(obj.encodePrettily());
-                      stream.resume();
-                    });
-                if ((cnt.get() % 1000) == 0) {
-                  log.info("cnt = {}", cnt);
-                }
-              });
-              stream.endHandler(end -> {
-                tx.commit().compose(y -> connection.close());
-                response.write("\n]\n}");
-                response.end();
-              });
-              stream.exceptionHandler(e -> {
-                log.error("stream error {}", e.getMessage(), e);
-                tx.commit().compose(y -> connection.close());
-              });
-              return Future.succeededFuture();
-            })));
+  public Future<Void> getClusters(RoutingContext ctx, String matchKeyId,
+      String sqlWhere, String sqlOrderBy) {
+    String from = clusterRecordTable + " LEFT JOIN " + clusterValueTable + " ON "
+        + clusterValueTable + ".cluster_id = " + clusterRecordTable + ".cluster_id"
+        + " WHERE " + clusterRecordTable + ".match_key_config_id = " + postgresString(matchKeyId);
```

Comment:
Is 'postgresString' sufficient to avoid injection?

Reply:
No. I should use parameterized queries for this or make this one better :-)

Reply:
Using parameterized query now